### PR TITLE
resource_manager/client: fix exceed error when ru consump of query is greater than ru fill rate

### LINKED ISSUE #6298: Exceeded resource group quota limitation when ru consump of query is greater than ru fill rate
### What did you do?
use `ycsb` insert test.  The table has 100 fileds which is 3000 bytes.
Resource group config : fill_rate=100，burst_limit=50M, token = 50M

### What did you expect to see?
Run test successfully.

### What did you see instead?
Exceeded resource group quota limitation

### What version of PD are you using (`pd-server -V`)?
master

### LINKED ISSUE #6300: Exceeded resource group quota limitation when waiting request too much
### What did you do?
use ycsb insert test. The table has 100 fileds which is 3000 bytes.
Resource group config : fill_rate=100，burst_limit=50M, token = 50M

### What did you expect to see?
Run test successfully.

### What did you see instead?
Exceeded resource group quota limitation
Output debug log
> [2023/04/11 07:18:38.282 +00:00] [INFO] [limiter.go:318] ["[resource group controller] after reconfigure"] [NewTokens=-471.16899387920347] [NewRate=110.37097769444105] [NotifyThreshold=0] [burst=5000000]
[2023/04/11 07:18:42.491 +00:00] [INFO] [limiter.go:302] ["[resource group controller] before reconfigure"] [NewTokens=-9.940384568614341] [NewRate=110.37097769444105] [NotifyThreshold=0] [burst=5000000]
[2023/04/11 07:18:42.491 +00:00] [INFO] [limiter.go:318] ["[resource group controller] after reconfigure"] [NewTokens=-9.905865052001605] [NewRate=20.57382465462127] [NotifyThreshold=0] [burst=5000000]
[2023/04/11 07:18:43.629 +00:00] [WARN] [limiter.go:377] ["[resource group controller]"] [Tokens=-592.8721281986398] [Rate=20.57382465462127] [n=99.5498046875] [burst=5000000] [low_ru=false] [notifythrelod=0]

By simply calculating the number of tokens, we can confirm that there are no new requests during this period, so client is waiting for the previous requests to be digested. This leads to a serious problem. There is no additional RU consumption in this resource group, so the consumption speed of RU slows down. As a result, when obtaining tokens from the server, there are very few tokens, but then KV request is generated again in large numbers, exceeded resource group quota limitation

### What version of PD are you using (`pd-server -V`)?
master

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?

Issue Number: Close #6298 close #6300

### What is changed and how does it work?
1. notify when the limiter reserve token fails. In addition, the limit of notify times is set when the limiter is limit state.
2. Update the logic of RU consumption statistics. This ensures that speed includes the application in progress when notifying

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

### Release note

```release-note
None.
```

### CODE REVIEW

File: client/resource_group/controller/controller.go [hunk 1]
```diff
@@ -1019,18 +1023,18 @@ func (gc *groupCostController) onRequestWait(
 					break retryLoop
 				}
 			}
-			time.Sleep(100 * time.Millisecond)
+			time.Sleep(retryInterval)
```
Comment (nolouch) [inline client/resource_group/controller/controller.go:1026]: Better add a metrics.

File: client/resource_group/controller/controller.go [hunk 2]
```diff
@@ -990,6 +991,9 @@ func (gc *groupCostController) onRequestWait(
 	for _, calc := range gc.calculators {
 		calc.BeforeKVRequest(delta, info)
 	}
+	gc.mu.Lock()
```
Comment (nolouch) [inline client/resource_group/controller/controller.go:994]: I wonder it's hurt the performance, better move out for this hot path.

File: client/resource_group/controller/limiter.go [hunk 3]
```diff
@@ -370,6 +370,10 @@ func (lim *Limiter) reserveN(now time.Time, n float64, maxFutureReserve time.Dur
 		lim.maybeNotify()
 	} else {
 		lim.last = last
+		if lim.limit == 0 {
```
Comment (nolouch) [inline client/resource_group/controller/limiter.go:373]: why need `limit == 0`?

Comment (CabinfeverB) [reply client/resource_group/controller/limiter.go:373]: When fillrate is greater than 0, it is in trickle state or degraded mode. If you are in trickle state, the server is allocating tokens as much as possible. You don't need to take tokens.

File: client/resource_group/controller/limiter.go [hunk 4]
```diff
@@ -370,10 +372,27 @@ func (lim *Limiter) reserveN(now time.Time, n float64, maxFutureReserve time.Dur
...
+			lim.notify()
+		} else {
+			// When fillrate is greater than 0, the speed limit is already set.
+			// If limiter are in limit state, the server has allocated tokens as much as possible. Don't need to request tokens.
+			if lim.notifyTime > 0 {
+				lim.notifyTime--
+				lim.notify()
+			}
+		}
+		log.Warn("[resource group controller]", zap.Float64("NewTokens", lim.tokens), zap.Float64("NewRate", float64(lim.limit)), zap.Float64("n", n), zap.Int64("burst", lim.burst))
```
Comment (nolouch) [inline client/resource_group/controller/limiter.go:385]:
```suggestion
		log.Warn("[resource group controller]", zap.Float64("current tokens", lim.tokens), zap.Float64("current rate", float64(lim.limit)), zap.Float64("request tokens", n), zap.Int64("burst", lim.burst))
```

Comment (nolouch) [reply client/resource_group/controller/limiter.go:385]: add add the notifytimes as a filed in the log.

File: client/resource_group/controller/controller.go [hunk 5]
```diff
@@ -345,6 +349,14 @@ func (c *ResourceGroupsController) updateAvgRequestResourcePerSec() {
...
 }
 
+func (c *ResourceGroupsController) resetEmergencyTokenAcquisition() {
+	c.groupsController.Range(func(name, value any) bool {
+		gc := value.(*groupCostController)
+		gc.resetEmergencyTokenAcquisition()
+		return true
+	})
+}
+
```
Comment (HuSharp) [inline client/resource_group/controller/controller.go:359]: for all including `c.groupsController.Range` funcs
Do you think it needs to abstract out all the traverse ranges and pass in specific functions each time?

Comment (CabinfeverB) [reply client/resource_group/controller/controller.go:359]: Good idea. We can impl it in another PR

Comment (HuSharp) [reply client/resource_group/controller/controller.go:359]: Ok, I will try.

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**